spmd/compiler/fusion.py

```diff
@@ -32,25 +37,45 @@ class FusionElement:
     """This class tracks the nodes for a DTensor expanded communication collective in the graph"""
 
     in_graph: bool = False
+    processed: bool = False
+    size: Optional[int] = 0
+    shape: Optional[Iterable[[int], [int]]] = field(default_factory=lambda: [])  # type: ignore
     comm_type: Optional[CommType] = None
     node_list: Optional[List[fx.Node]] = field(default_factory=lambda: [])  # type: ignore
-    size: int = 0
     prev_node: Optional[fx.Node] = None  # node that was before start of section
-    next_node: Optional[fx.Node] = None  # node that was after end of section
-    processed: bool = False
     output_name: str = ""
-    clone_node: Optional[fx.Node] = None
-    comm_node: Optional[fx.Node] = None
-    wait_node: Optional[fx.Node] = None
+    comm_node: Optional[fx.Node] = None  # type: ignore
+    wait_node: Optional[fx.Node] = None  # type: ignore
+    grad_tensor_node: Optional[fx.Node] = None  # type: ignore
+
+    def _get_next_node(
+        self,
+    ) -> fx.Node:
+        """get the next node after this FE section"""
+        next_node = self.node_list[-1].next  # type: ignore
+        # _debug(f"57, next node name is {next_node.name}")
+        assert (
+            next_node is not None
+        ), f"failed to get valid next node after {self.node_list[-1].name}"  # type: ignore
+        return next_node
 
 
 @dataclass
 class GraphInfo:
+    """provides a home for global aspects of this graph.
+    Currently tracks first and last node, len of the graph and
```

Comment:
What does "first" and "last" node mean?

Reply:
this is the very first and very last node of the graph.  
Atm, fx.graph does not keep/expose a tail, so I wanted to be able to store and access front and back since we have to do a full graph scan to start.  (you can reverse iter but it's not clear that internally it isn't running the entire graph to do that first, which could be slow for larger graphs).

Reply:
I had the same question in the doc. Can you mention this in your comment as well?